SU2_CFD/src/drivers/CDiscAdjMultizoneDriver.cpp

```diff
@@ -441,47 +440,21 @@ void CDiscAdjMultizoneDriver::SetRecording(unsigned short kind_recording, Kind_T
 
   AD::Push_TapePosition();
 
-  /*--- Extract the objective function and store it --- */
-  
-  for(iZone = 0; iZone < nZone; iZone++) {
+  /*--- Extract the objective function and store it.
+   *    It is necessary to include data transfer and mesh updates in this section as some functions
+   *    computed in one zone depend explicitly on the variables of others through that path. --- */
 
-    /*--- In principle, the mesh does not need to be updated ---*/
-    bool DeformMesh = false;
-
-    /*--- Transfer from all the remaining zones ---*/
-    for (jZone = 0; jZone < nZone; jZone++){
-      /*--- The target zone is iZone ---*/
-      if (jZone != iZone && interface_container[iZone][jZone] != NULL) {
-        DeformMesh = DeformMesh || Transfer_Data(jZone, iZone);
-      }
-    }
-    /*--- If a mesh update is required due to the transfer of data ---*/
-    if (DeformMesh) DynamicMeshUpdate(iZone, ExtIter);
-  }
+  HandleDataTransfer();
```

Comment:
@oleburghardt @rsanfer @cvencro 
FSI adjoints seem to be working with the new driver but with legacy deformation and transfer classes (at least the regression case works in serial, needs more testing).
If you want to try it out you need to change line 127 of SU2_CFD.cpp to instantiate the new driver, the "legacy" ADJ_FSI option still needs to be set.

@oleburghardt one key change is this one I highlight here, the transfer and deformation needs to be included also in the OF part of the tape as since the mesh deformation is not treated as a third zone some fluid objectives depend explicitly on structural variables through the transfer+deformation path.

Reply:
@pcarruscag I've tried with the new driver instantiated for FSI adjoints, gradients from both serial and parallel runs are accurate.